mentat/llm_api_handler.py

```diff
@@ -134,6 +137,74 @@ def count_tokens(message: str, model: str, full_message: bool) -> int:
     )
 
 
+def normalize_messages_for_anthropic(
+    messages: list[ChatCompletionMessageParam],
+) -> list[ChatCompletionMessageParam]:
+    """Claude expects the chat to start with at most one system message and afterwards user and system messages to
+    alternate. This method consolidates all the system messages at the beginning of the conversation into one system
+    message delimited by "\n"+"-"*80+"\n and turns future system messages into user messages annotated with "System:"
+    and combines adjacent assistant or user messages into one assistant or user message.
+    """
+    replace_non_leading_systems = list[ChatCompletionMessageParam]()
+    for i, message in enumerate(messages):
+        if message["role"] == "system":
+            if i == 0 or messages[i - 1]["role"] == "system":
+                replace_non_leading_systems.append(message)
+            else:
+                content = "SYSTEM: " + (message["content"] or "")
+                replace_non_leading_systems.append(
+                    ChatCompletionUserMessageParam(role="user", content=content)
+                )
+        else:
+            replace_non_leading_systems.append(message)
+
+    concatenate_adjacent = list[ChatCompletionMessageParam]()
+    current_role: str = ""
+    current_content: str = ""
+    delimiter = "\n" + "-" * 80 + "\n"
+    for message in replace_non_leading_systems:
+        if message["role"] == current_role:
+            current_content += delimiter + str(message["content"])
+        else:
+            if current_role == "user":
+                concatenate_adjacent.append(
+                    ChatCompletionUserMessageParam(
+                        role=current_role, content=current_content
+                    )
+                )
+            elif current_role == "system":
+                concatenate_adjacent.append(
+                    ChatCompletionSystemMessageParam(
+                        role=current_role, content=current_content
+                    )
+                )
+            elif current_role == "assistant":
+                concatenate_adjacent.append(
+                    ChatCompletionAssistantMessageParam(
+                        role=current_role, content=current_content
+                    )
+                )
+            current_role = message["role"]
+            current_content = str(message["content"])
+
+    if current_role == "user":
+        concatenate_adjacent.append(
+            ChatCompletionUserMessageParam(role=current_role, content=current_content)
+        )
+    elif current_role == "system":
+        concatenate_adjacent.append(
+            ChatCompletionSystemMessageParam(role=current_role, content=current_content)
+        )
+    elif current_role == "assistant":
+        concatenate_adjacent.append(
+            ChatCompletionAssistantMessageParam(
+                role=current_role, content=current_content
+            )
+        )
```

Comment:
@abanteai-butler can you refactor, putting the if/elif/elif to append messages into a subfunction that's called in the loop and after the loop ends?

Reply:
please write the suggestion in github suggestion syntax

Reply:
you have to rewrite everything, can't just say `# Then replace the if/elif/elif blocks in the loop and after the loop with calls to this function.` or when I accept the suggestion it will overwrite everything

Reply:
not quite, the suggestion is just replacing the code starting at the line `concatenate_adjacent = list[ChatCompletionMessageParam]()`, so you don't need to include the code from before that line, that will remain

Reply:
you didn't actually make the suggestion